stdlib/runtime/runtimeDebug.gr

```diff
@@ -0,0 +1,712 @@
+import GC from "runtime/gc"
+import { coerceNumberToWasmI32 } from "runtime/numbers"
+import { allocateString } from "runtime/dataStructures"
+import NumberUtils from "runtime/numberUtils"
+import Map from "map"
+import Set from "set"
+import Option from "option"
+import Tags from "runtime/unsafe/tags"
+
+@disableGC
+let rec toHex = (n: Number) => {
+  let x = coerceNumberToWasmI32(n)
+  let result = NumberUtils.itoa32(x, 16n)
+
+  Memory.decRef(WasmI32.fromGrain(n))
+  Memory.decRef(WasmI32.fromGrain(toHex))
+
+  result
+}
+
+export enum LeakDebugSummaryMode {
+  AlwaysPrintSummary,
+  PrintSummaryIfLeaksAreFound,
+}
+
+export record LeakDebugBlockOptions {
+  printIncrefsAndDecRefs: Bool,
+  printMallocAndFree: Bool,
+  printObjectContents: Bool,
+  summary: LeakDebugSummaryMode,
+}
+
+let incrementCountInMap = (key: k, map: Map.Map<k, Number>) => {
+  match (Map.get(key, map)) {
+    Some(cnt) => Map.set(key, cnt + 1, map),
+    None => Map.set(key, 1, map),
+  }
+}
+
+let keyUnionOfTwoMaps = (map1: Map.Map<k, v1>, map2: Map.Map<k, v2>) => {
+  let keyUnion = Set.make(): (Set.Set<k>)
+
+  Map.forEach(
+    (k, v) => {
+      Set.add(k, keyUnion)
+    },
+    map1,
+  )
+
+  Map.forEach(
+    (k, v) => {
+      Set.add(k, keyUnion)
+    },
+    map2,
+  )
+
+  keyUnion
+}
+
+let leakDebugBlockImpl = (options: LeakDebugBlockOptions, f: () -> Void) => {
+  // The main objective is to find and print objects allocated
+  // in the function, but not deallocated at the end.
+
+  let allocations = Set.make(): (Set.Set<Number>)
+
+  let tagToTypeName = (tag: Number) => {
+    match (tag) {
+      1 => "string",
+      2 => "char",
+      3 => "enum",
+      4 => "record",
+      5 => "array",
+      6 => "number",
+      7 => "lambda",
+      8 => "tuple",
+      9 => "bytes",
+      _ => "unknown",
+    }
+  }
+
+  let mut totalMallocs = 0
+  let mut totalFrees = 0
+  let mut totalIncRefs = 0
+  let mut totalDecRefs = 0
+
+  let mut totalPreexistingObjectsDeallocated = 0
+
+  let incRefsByTag = Map.make(): (Map.Map<Number, Number>)
+  let decRefsByTag = Map.make(): (Map.Map<Number, Number>)
+
+  let inintialRefCountsOnPreExistingObjects = Map.make(): (Map.Map<Number, Number>)
+  let incRefsOnPreExistingObjects = Map.make(): (Map.Map<Number, Number>)
+  let decRefsOnPreExistingObjects = Map.make(): (Map.Map<Number, Number>)
+
+  // Since objects created by GC.malloc have a starting refcount of 1,
+  // the total incRefsByTag would not match decRefsByTag. I'm compensating it
+  // by keeping track of number of objects that have gone malloc+free cycle.
+  // This isn't perfect because in case of leaks free isn't called and the
+  // totals can create confusion. The simplest solution would be to get the tag
+  // of mallocated objects, but they don't yet have a tag associated. So I
+  // have to compensate again by adding up the objects that have been
+  // mallocated but never freed. Which simply means the ones in "allocations"
+  // Map.
```

Comment:
This part loses me a little. Can you maybe elaborate a bit?

Reply:
I agree this comment is pretty confusing. I'll rewrite it.

What I mean is that to correctly present the summary of incRefs/decRefs by tag, it's not enough to just track incRefs/decRefs themselves because incRef doesn't get called for newly allocated objects. They start with ref. count of 1 set directly in `GC.malloc`. But it also isn't as simple as adding some code in the hook for `GC.malloc` because freshly allocated objects don't have a tag yet. The tag is normally set after (like in `dataStructures.allocateString` and similar functions). So what I actually do is track the lifetime of single objects in the `allocations` map. Only when the object is freed I add to the `mallocPlusFreeByTag` map to present along with incRefs/decRefs by tag. If an object is not freed then it's left in the `allocations` map. So I group these leaked objects by tag as well and present the totals in the summary as "malloc without free".